File: 02- February/25- Best Time to Buy and Sell Stock/25- Best Time to Buy and Sell Stock (Noura Algohary).cpp
// Author: Noura Algohary
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int left = prices[0], right = prices[0], profit = 0, max_profit = 0;

        for(int price: prices)
        {
            // a new min point needs a new max point in the future
            if(price < left)
                {
                left = price;
                right = price;
                }
            else if(right < price)
            {
                right = price;  // record another max for the same min point
                profit = right - left;
            }
            if(profit > max_profit)
                max_profit = profit;
        }

        return max_profit;
    }
};

File: 02- February/25- Best Time to Buy and Sell Stock/25- Best Time to Buy and Sell Stock (Noura Algohary).py
# Author: Noura Algohary
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left, right = prices[0], prices[0]
        profit, max_profit = 0, 0

        for price in prices:
            
            # a new min point needs a new max point in the future
            if price < left:
                left = price    
                right = price
            elif price > right:
                right = price   # record another max for the same min point
                profit = right - left   

                if profit > max_profit:
                    max_profit = profit
        
        return max_profit